File: TypeScript/__tests__/ExpressionKit.test.ts
/*
 * MIT License
 *
 * Copyright (c) 2025 ExpressionKit Contributors
 *
 * ExpressionKit TypeScript Test Suite
 * 
 * This test suite ensures 1:1 behavioral parity with the C++ and Swift implementations.
 * Test cases are translated from the C++ test.cpp file to ensure identical behavior.
 */

import {
    Expression,
    Value,
    ValueType,
    ExpressionError,
    SimpleEnvironment,
    TokenType,
    callStandardFunctions
} from '../src/index';

describe('ExpressionKit TypeScript Tests', () => {
    describe('Value Tests', () => {
        test('Number Value Creation and Conversion', () => {
            const value = new Value(42);
            expect(value.type).toBe(ValueType.NUMBER);
            expect(value.asNumber()).toBe(42);
            expect(value.asBoolean()).toBe(true);
            expect(value.asString()).toBe('42');
        });

        test('Boolean Value Creation and Conversion', () => {
            const trueValue = new Value(true);
            expect(trueValue.type).toBe(ValueType.BOOLEAN);
            expect(trueValue.asNumber()).toBe(1);
            expect(trueValue.asBoolean()).toBe(true);
            expect(trueValue.asString()).toBe('true');

            const falseValue = new Value(false);
            expect(falseValue.type).toBe(ValueType.BOOLEAN);
            expect(falseValue.asNumber()).toBe(0);
            expect(falseValue.asBoolean()).toBe(false);
            expect(falseValue.asString()).toBe('false');
        });

        test('String Value Creation and Conversion', () => {
            const value = new Value('hello');
            expect(value.type).toBe(ValueType.STRING);
            expect(value.asString()).toBe('hello');
            expect(value.asBoolean()).toBe(true);

            const emptyString = new Value('');
            expect(emptyString.asBoolean()).toBe(false);

            const numericString = new Value('123.45');
            expect(numericString.asNumber()).toBeCloseTo(123.45);

            expect(() => new Value('not_a_number').asNumber()).toThrow(ExpressionError);
        });

        test('Value Equality', () => {
            expect(new Value(42).equals(new Value(42))).toBe(true);
            expect(new Value(true).equals(new Value(true))).toBe(true);
            expect(new Value('hello').equals(new Value('hello'))).toBe(true);
            
            expect(new Value(42).equals(new Value(43))).toBe(false);
            expect(new Value(42).equals(new Value(true))).toBe(false);
        });
    });

    describe('Simple Environment Tests', () => {
        test('Variable Storage and Retrieval', () => {
            const env = new SimpleEnvironment();
            env.setVariable('x', new Value(10));
            env.setVariable('y', new Value(20));
            
            expect(env.get('x').asNumber()).toBe(10);
            expect(env.get('y').asNumber()).toBe(20);
            expect(() => env.get('unknown')).toThrow(ExpressionError);
        });

        test('Standard Mathematical Functions', () => {
            const env = new SimpleEnvironment();
            
            // Test min/max
            expect(env.call('min', [new Value(5), new Value(10)]).asNumber()).toBe(5);
            expect(env.call('max', [new Value(5), new Value(10)]).asNumber()).toBe(10);
            
            // Test mathematical functions
            expect(env.call('sqrt', [new Value(16)]).asNumber()).toBe(4);
            expect(env.call('abs', [new Value(-5)]).asNumber()).toBe(5);
            expect(env.call('pow', [new Value(2), new Value(3)]).asNumber()).toBe(8);
            
            // Test domain errors
            expect(() => env.call('unknown_function', [])).toThrow(ExpressionError);
        });
    });

    describe('Basic Expression Evaluation', () => {
        test('Simple Arithmetic', () => {
            expect(Expression.evalSimple('2 + 3').asNumber()).toBe(5);
            expect(Expression.evalSimple('10 - 4').asNumber()).toBe(6);
            expect(Expression.evalSimple('3 * 4').asNumber()).toBe(12);
            expect(Expression.evalSimple('15 / 3').asNumber()).toBe(5);
        });

        test('Operator Precedence', () => {
            expect(Expression.evalSimple('2 + 3 * 4').asNumber()).toBe(14);
            expect(Expression.evalSimple('(2 + 3) * 4').asNumber()).toBe(20);
            expect(Expression.evalSimple('2 * 3 + 4').asNumber()).toBe(10);
            expect(Expression.evalSimple('2 * (3 + 4)').asNumber()).toBe(14);
        });

        test('Boolean Literals', () => {
            expect(Expression.evalSimple('true').asBoolean()).toBe(true);
            expect(Expression.evalSimple('false').asBoolean()).toBe(false);
        });

        test('String Literals', () => {
            expect(Expression.evalSimple('"hello"').asString()).toBe('hello');
            expect(Expression.evalSimple('"Hello, World!"').asString()).toBe('Hello, World!');
            expect(Expression.evalSimple('""').asString()).toBe('');
        });

        test('Unary Operators', () => {
            expect(Expression.evalSimple('-5').asNumber()).toBe(-5);
            expect(Expression.evalSimple('!true').asBoolean()).toBe(false);
            expect(Expression.evalSimple('!false').asBoolean()).toBe(true);
            expect(Expression.evalSimple('not true').asBoolean()).toBe(false);
        });
    });

    describe('Comparison Operations', () => {
        test('Relational Operators', () => {
            expect(Expression.evalSimple('5 < 10').asBoolean()).toBe(true);
            expect(Expression.evalSimple('10 < 5').asBoolean()).toBe(false);
            expect(Expression.evalSimple('5 > 10').asBoolean()).toBe(false);
            expect(Expression.evalSimple('10 > 5').asBoolean()).toBe(true);
            expect(Expression.evalSimple('5 <= 5').asBoolean()).toBe(true);
            expect(Expression.evalSimple('5 >= 5').asBoolean()).toBe(true);
        });

        test('Equality Operators', () => {
            expect(Expression.evalSimple('5 == 5').asBoolean()).toBe(true);
            expect(Expression.evalSimple('5 == 6').asBoolean()).toBe(false);
            expect(Expression.evalSimple('5 != 6').asBoolean()).toBe(true);
            expect(Expression.evalSimple('5 != 5').asBoolean()).toBe(false);
        });
    });

    describe('Logical Operations', () => {
        test('AND Operations', () => {
            expect(Expression.evalSimple('true && true').asBoolean()).toBe(true);
            expect(Expression.evalSimple('true && false').asBoolean()).toBe(false);
            expect(Expression.evalSimple('false && true').asBoolean()).toBe(false);
            expect(Expression.evalSimple('false && false').asBoolean()).toBe(false);
            expect(Expression.evalSimple('true and true').asBoolean()).toBe(true);
        });

        test('OR Operations', () => {
            expect(Expression.evalSimple('true || false').asBoolean()).toBe(true);
            expect(Expression.evalSimple('false || true').asBoolean()).toBe(true);
            expect(Expression.evalSimple('false || false').asBoolean()).toBe(false);
            expect(Expression.evalSimple('true || true').asBoolean()).toBe(true);
            expect(Expression.evalSimple('true or false').asBoolean()).toBe(true);
        });

        test('XOR Operations', () => {
            expect(Expression.evalSimple('true xor false').asBoolean()).toBe(true);
            expect(Expression.evalSimple('false xor true').asBoolean()).toBe(true);
            expect(Expression.evalSimple('true xor true').asBoolean()).toBe(false);
            expect(Expression.evalSimple('false xor false').asBoolean()).toBe(false);
        });

        test('Short-Circuit Evaluation', () => {
            const env = new SimpleEnvironment();
            env.setVariable('x', new Value(0));
            
            // This should not cause division by zero due to short-circuit evaluation
            expect(Expression.evalSimple('false && (1 / x) > 0', env).asBoolean()).toBe(false);
            expect(Expression.evalSimple('true || (1 / x) > 0', env).asBoolean()).toBe(true);
        });
    });

    describe('Variable Access', () => {
        test('Simple Variables', () => {
            const env = new SimpleEnvironment();
            env.setVariable('x', new Value(10));
            env.setVariable('y', new Value(20));
            
            expect(Expression.evalSimple('x', env).asNumber()).toBe(10);
            expect(Expression.evalSimple('y', env).asNumber()).toBe(20);
            expect(Expression.evalSimple('x + y', env).asNumber()).toBe(30);
        });

        test('Dot Notation Variables', () => {
            const env = new SimpleEnvironment();
            env.setVariable('pos.x', new Value(10.5));
            env.setVariable('pos.y', new Value(20.3));
            
            expect(Expression.evalSimple('pos.x', env).asNumber()).toBeCloseTo(10.5);
            expect(Expression.evalSimple('pos.y', env).asNumber()).toBeCloseTo(20.3);
        });

        test('Unknown Variable Error', () => {
            const env = new SimpleEnvironment();
            expect(() => Expression.evalSimple('unknown_var', env)).toThrow(ExpressionError);
        });
    });

    describe('Function Calls', () => {
        test('Standard Mathematical Functions', () => {
            expect(Expression.evalSimple('min(5, 10)').asNumber()).toBe(5);
            expect(Expression.evalSimple('max(5, 10)').asNumber()).toBe(10);
            expect(Expression.evalSimple('sqrt(16)').asNumber()).toBe(4);
            expect(Expression.evalSimple('abs(-5)').asNumber()).toBe(5);
            expect(Expression.evalSimple('pow(2, 3)').asNumber()).toBe(8);
            expect(Expression.evalSimple('floor(3.7)').asNumber()).toBe(3);
            expect(Expression.evalSimple('ceil(3.2)').asNumber()).toBe(4);
            expect(Expression.evalSimple('round(3.6)').asNumber()).toBe(4);
        });

        test('Trigonometric Functions', () => {
            expect(Expression.evalSimple('sin(0)').asNumber()).toBeCloseTo(0);
            expect(Expression.evalSimple('cos(0)').asNumber()).toBeCloseTo(1);
            expect(Expression.evalSimple('tan(0)').asNumber()).toBeCloseTo(0);
        });

        test('Logarithmic Functions', () => {
            expect(Expression.evalSimple('exp(0)').asNumber()).toBeCloseTo(1);
            expect(Expression.evalSimple('log(1)').asNumber()).toBeCloseTo(0);
        });

        test('Function Domain Errors', () => {
            // sqrt of negative number should fail gracefully
            expect(() => Expression.evalSimple('sqrt(-1)')).toThrow();
            // log of non-positive number should fail gracefully  
            expect(() => Expression.evalSimple('log(0)')).toThrow();
            expect(() => Expression.evalSimple('log(-1)')).toThrow();
        });
    });

    describe('Complex Expressions', () => {
        test('Nested Function Calls', () => {
            expect(Expression.evalSimple('max(min(10, 5), 3)').asNumber()).toBe(5);
            expect(Expression.evalSimple('sqrt(pow(3, 2) + pow(4, 2))').asNumber()).toBe(5);
        });

        test('Mixed Operations', () => {
            const env = new SimpleEnvironment();
            env.setVariable('health', new Value(100));
            env.setVariable('maxHealth', new Value(100));
            env.setVariable('level', new Value(5));
            
            expect(Expression.evalSimple('health / maxHealth', env).asNumber()).toBe(1);
            expect(Expression.evalSimple('health < maxHealth * 0.5', env).asBoolean()).toBe(false);
            expect(Expression.evalSimple('level > 0 && health > 0', env).asBoolean()).toBe(true);
        });
    });

    describe('Error Handling', () => {
        test('Division by Zero', () => {
            expect(() => Expression.evalSimple('1 / 0')).toThrow(ExpressionError);
            expect(() => Expression.evalSimple('10 / (5 - 5)')).toThrow(ExpressionError);
        });

        test('Parse Errors', () => {
            expect(() => Expression.evalSimple('2 +')).toThrow(ExpressionError);
            expect(() => Expression.evalSimple('2 * * 3')).toThrow(ExpressionError);
            expect(() => Expression.evalSimple('(2 + 3')).toThrow(ExpressionError);
            expect(() => Expression.evalSimple(')')).toThrow(ExpressionError);
        });

        test('Unknown Function', () => {
            expect(() => Expression.evalSimple('unknown_function(1, 2)')).toThrow(ExpressionError);
        });
    });

    describe('Pre-compiled Expressions', () => {
        test('Parse Once, Execute Many', () => {
            const env = new SimpleEnvironment();
            const compiled = Expression.parse('x * 2 + y');
            
            env.setVariable('x', new Value(5));
            env.setVariable('y', new Value(3));
            expect(compiled.evaluate(env).asNumber()).toBe(13);
            
            env.setVariable('x', new Value(10));
            env.setVariable('y', new Value(7));
            expect(compiled.evaluate(env).asNumber()).toBe(27);
        });

        test('Performance Benefits', () => {
            // This test demonstrates the performance benefit of pre-compilation
            // but doesn't actually measure performance in the unit test
            const env = new SimpleEnvironment();
            env.setVariable('x', new Value(5));
            
            const compiled = Expression.parse('x * 2 + sqrt(x) - floor(x / 2)');
            
            for (let i = 0; i < 100; i++) {
                env.setVariable('x', new Value(i));
                const result = compiled.evaluate(env);
                expect(result.type).toBe(ValueType.NUMBER);
            }
        });
    });

    describe('Token Collection', () => {
        test('Basic Token Collection', () => {
            const [result, tokens] = Expression.eval('2 + 3', undefined, true);
            
            expect(result.asNumber()).toBe(5);
            expect(tokens).not.toBeNull();
            expect(tokens!.length).toBeGreaterThan(0);
            
            // Check that we have the expected token types
            const tokenTypes = tokens!.map(t => t.type);
            expect(tokenTypes).toContain(TokenType.NUMBER);
            expect(tokenTypes).toContain(TokenType.OPERATOR);
        });

        test('Function Call Token Collection', () => {
            const [result, tokens] = Expression.eval('max(10, 5)', undefined, true);
            
            expect(result.asNumber()).toBe(10);
            expect(tokens).not.toBeNull();
            
            const tokenTypes = tokens!.map(t => t.type);
            expect(tokenTypes).toContain(TokenType.IDENTIFIER);
            expect(tokenTypes).toContain(TokenType.PARENTHESIS);
            expect(tokenTypes).toContain(TokenType.NUMBER);
            expect(tokenTypes).toContain(TokenType.COMMA);
        });

        test('String Token Collection', () => {
            const [result, tokens] = Expression.eval('"hello world"', undefined, true);
            
            expect(result.asString()).toBe('hello world');
            expect(tokens).not.toBeNull();
            
            const tokenTypes = tokens!.map(t => t.type);
            expect(tokenTypes).toContain(TokenType.STRING);
        });

        test('No Token Collection', () => {
            const [result, tokens] = Expression.eval('2 + 3', undefined, false);
            
            expect(result.asNumber()).toBe(5);
            expect(tokens).toBeNull();
        });
    });

    describe('String Literal Parsing', () => {
        test('Basic String Literals', () => {
            expect(Expression.evalSimple('"hello"').asString()).toBe('hello');
            expect(Expression.evalSimple('""').asString()).toBe('');
            expect(Expression.evalSimple('"Hello, World!"').asString()).toBe('Hello, World!');
        });

        test('Escaped Characters', () => {
            expect(Expression.evalSimple('"hello\\nworld"').asString()).toBe('hello\nworld');
            expect(Expression.evalSimple('"hello\\tworld"').asString()).toBe('hello\tworld');
            expect(Expression.evalSimple('"hello\\\\world"').asString()).toBe('hello\\world');
            expect(Expression.evalSimple('"say \\"hello\\""').asString()).toBe('say "hello"');
        });

        test('Unterminated String Error', () => {
            expect(() => Expression.evalSimple('"unterminated')).toThrow(ExpressionError);
        });
    });

    describe('Edge Cases and Compatibility', () => {
        test('Number Parsing Edge Cases', () => {
            expect(Expression.evalSimple('0').asNumber()).toBe(0);
            expect(Expression.evalSimple('0.0').asNumber()).toBe(0);
            expect(Expression.evalSimple('.5').asNumber()).toBe(0.5);
            expect(Expression.evalSimple('123.456').asNumber()).toBeCloseTo(123.456);
        });

        test('Boolean Conversion Edge Cases', () => {
            expect(new Value(0).asBoolean()).toBe(false);
            expect(new Value(0.0).asBoolean()).toBe(false);
            expect(new Value(1).asBoolean()).toBe(true);
            expect(new Value(-1).asBoolean()).toBe(true);
            expect(new Value(0.1).asBoolean()).toBe(true);
        });

        test('Expression Without Environment', () => {
            // Should work for constants and standard functions
            expect(Expression.evalSimple('2 + 3').asNumber()).toBe(5);
            expect(Expression.evalSimple('sqrt(16)').asNumber()).toBe(4);
            expect(Expression.evalSimple('true && false').asBoolean()).toBe(false);
            
            // Should fail for variables
            expect(() => Expression.evalSimple('x + 1')).toThrow(ExpressionError);
        });
    });
});

File: Kotlin/src/test/kotlin/ExpressionKitTest.kt
/*
 * MIT License
 *
 * Copyright (c) 2025 ExpressionKit Contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This test suite maintains 1:1 correspondence with the C++ and Swift test suites
 * to ensure behavioral consistency across all language implementations.
 */

package com.expressionkit

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.BeforeEach
import kotlin.math.*

class ExpressionKitTest {
    
    private lateinit var env: SimpleEnvironment
    
    @BeforeEach
    fun setup() {
        env = SimpleEnvironment()
    }
    
    // MARK: - Basic Arithmetic Tests
    
    @Test
    fun testBasicArithmetic() {
        assertEquals(Value.number(5.0), Expression.eval("2 + 3"))
        assertEquals(Value.number(7.0), Expression.eval("10 - 3"))
        assertEquals(Value.number(20.0), Expression.eval("4 * 5"))
        assertEquals(Value.number(5.0), Expression.eval("15 / 3"))
        assertEquals(Value.number(14.0), Expression.eval("2 + 3 * 4"))
    }
    
    @Test
    fun testArithmeticPrecedence() {
        assertEquals(Value.number(14.0), Expression.eval("2 + 3 * 4")) // 2 + (3 * 4)
        assertEquals(Value.number(10.0), Expression.eval("2 * 3 + 4")) // (2 * 3) + 4
        assertEquals(Value.number(4.0), Expression.eval("10 - 2 * 3")) // 10 - (2 * 3)
        assertEquals(Value.number(6.0), Expression.eval("12 / 3 + 2")) // (12 / 3) + 2
    }
    
    @Test
    fun testParenthesesGrouping() {
        assertEquals(Value.number(20.0), Expression.eval("(2 + 3) * 4"))
        assertEquals(Value.number(14.0), Expression.eval("2 * (3 + 4)"))
        assertEquals(Value.number(2.0), Expression.eval("(10 - 2) / (3 + 1)"))
        assertEquals(Value.number(19.0), Expression.eval("((2 + 3) * 4) - 1"))
    }
    
    @Test
    fun testComplexArithmetic() {
        assertEquals(Value.number(19.0), Expression.eval("(2 + 3) * 4 - 1"))
        assertEquals(Value.number(8.0), Expression.eval("10 / (2 + 3) * 4"))
        assertEquals(Value.number(11.0), Expression.eval("1 + 2 * 3 + 4"))
        assertEquals(Value.number(21.0), Expression.eval("(1 + 2) * (3 + 4)"))
    }
    
    // MARK: - Boolean Logic Tests
    
    @Test
    fun testBasicBooleanLogic() {
        assertEquals(Value.boolean(true), Expression.eval("true"))
        assertEquals(Value.boolean(false), Expression.eval("false"))
        assertEquals(Value.boolean(true), Expression.eval("true && true"))
        assertEquals(Value.boolean(false), Expression.eval("true && false"))
        assertEquals(Value.boolean(false), Expression.eval("false && false"))
    }
    
    @Test
    fun testLogicalOperators() {
        // AND operations
        assertEquals(Value.boolean(true), Expression.eval("true && true"))
        assertEquals(Value.boolean(false), Expression.eval("true && false"))
        assertEquals(Value.boolean(false), Expression.eval("false && true"))
        assertEquals(Value.boolean(false), Expression.eval("false && false"))
        
        // OR operations
        assertEquals(Value.boolean(true), Expression.eval("true || true"))
        assertEquals(Value.boolean(true), Expression.eval("true || false"))
        assertEquals(Value.boolean(true), Expression.eval("false || true"))
        assertEquals(Value.boolean(false), Expression.eval("false || false"))
        
        // NOT operations
        assertEquals(Value.boolean(false), Expression.eval("!true"))
        assertEquals(Value.boolean(true), Expression.eval("!false"))
        assertEquals(Value.boolean(false), Expression.eval("not true"))
        assertEquals(Value.boolean(true), Expression.eval("not false"))
    }
    
    @Test
    fun testTextualLogicalOperators() {
        // AND
        assertEquals(Value.boolean(true), Expression.eval("true and true"))
        assertEquals(Value.boolean(false), Expression.eval("true and false"))
        
        // OR  
        assertEquals(Value.boolean(true), Expression.eval("true or false"))
        assertEquals(Value.boolean(false), Expression.eval("false or false"))
        
        // NOT
        assertEquals(Value.boolean(false), Expression.eval("not true"))
        assertEquals(Value.boolean(true), Expression.eval("not false"))
    }
    
    @Test
    fun testLogicalPrecedence() {
        assertEquals(Value.boolean(true), Expression.eval("true || false && false")) // true || (false && false)
        assertEquals(Value.boolean(false), Expression.eval("false && true || false")) // (false && true) || false
        assertEquals(Value.boolean(true), Expression.eval("!false && true")) // (!false) && true
        assertEquals(Value.boolean(false), Expression.eval("!true || false")) // (!true) || false
    }
    
    // MARK: - Comparison Tests
    
    @Test
    fun testBasicComparisons() {
        assertEquals(Value.boolean(true), Expression.eval("5 > 3"))
        assertEquals(Value.boolean(false), Expression.eval("3 > 5"))
        assertEquals(Value.boolean(true), Expression.eval("3 < 5"))
        assertEquals(Value.boolean(false), Expression.eval("5 < 3"))
        assertEquals(Value.boolean(true), Expression.eval("5 >= 5"))
        assertEquals(Value.boolean(true), Expression.eval("5 >= 3"))
        assertEquals(Value.boolean(false), Expression.eval("3 >= 5"))
        assertEquals(Value.boolean(true), Expression.eval("3 <= 3"))
        assertEquals(Value.boolean(true), Expression.eval("3 <= 5"))
        assertEquals(Value.boolean(false), Expression.eval("5 <= 3"))
    }
    
    @Test
    fun testEqualityComparisons() {
        // Numbers
        assertEquals(Value.boolean(true), Expression.eval("5 == 5"))
        assertEquals(Value.boolean(false), Expression.eval("5 == 3"))
        assertEquals(Value.boolean(false), Expression.eval("5 != 5"))
        assertEquals(Value.boolean(true), Expression.eval("5 != 3"))
        
        // Booleans
        assertEquals(Value.boolean(true), Expression.eval("true == true"))
        assertEquals(Value.boolean(true), Expression.eval("false == false"))
        assertEquals(Value.boolean(false), Expression.eval("true == false"))
        assertEquals(Value.boolean(false), Expression.eval("false != false"))
        assertEquals(Value.boolean(true), Expression.eval("true != false"))
    }
    
    // MARK: - Variable Tests
    
    @Test
    fun testBasicVariables() {
        env.setVariable("x", 10.0)
        env.setVariable("y", 5.0)
        
        assertEquals(Value.number(10.0), Expression.eval("x", env))
        assertEquals(Value.number(5.0), Expression.eval("y", env))
        assertEquals(Value.number(15.0), Expression.eval("x + y", env))
        assertEquals(Value.number(50.0), Expression.eval("x * y", env))
    }
    
    @Test
    fun testBooleanVariables() {
        env.setVariable("isTrue", true)
        env.setVariable("isFalse", false)
        
        assertEquals(Value.boolean(true), Expression.eval("isTrue", env))
        assertEquals(Value.boolean(false), Expression.eval("isFalse", env))
        assertEquals(Value.boolean(false), Expression.eval("isTrue && isFalse", env))
        assertEquals(Value.boolean(true), Expression.eval("isTrue || isFalse", env))
    }
    
    @Test
    fun testStringVariables() {
        env.setVariable("name", "world")
        env.setVariable("greeting", "hello")
        
        assertEquals(Value.string("world"), Expression.eval("name", env))
        assertEquals(Value.string("hello"), Expression.eval("greeting", env))
        assertEquals(Value.string("helloworld"), Expression.eval("greeting + name", env))
    }
    
    @Test
    fun testVariableNotFound() {
        val exception = assertThrows<ExpressionError> {
            Expression.eval("unknownVar")
        }
        assertTrue(exception.message!!.contains("Variable 'unknownVar' not found"))
    }
    
    // MARK: - Function Tests
    
    @Test
    fun testBuiltinMathFunctions() {
        assertEquals(Value.number(4.0), Expression.eval("sqrt(16)"))
        assertEquals(Value.number(5.0), Expression.eval("abs(-5)"))
        assertEquals(Value.number(5.0), Expression.eval("abs(5)"))
        assertEquals(Value.number(8.0), Expression.eval("pow(2, 3)"))
        assertEquals(Value.number(3.0), Expression.eval("min(3, 5)"))
        assertEquals(Value.number(5.0), Expression.eval("max(3, 5)"))
    }
    
    @Test
    fun testTrigonometricFunctions() {
        // Test basic trig functions with known values
        val pi = PI
        assertEquals(0.0, Expression.eval("sin(0)").asNumber(), 1e-10)
        assertEquals(1.0, Expression.eval("cos(0)").asNumber(), 1e-10)
        assertEquals(0.0, Expression.eval("tan(0)").asNumber(), 1e-10)
    }
    
    @Test
    fun testLogarithmicFunctions() {
        assertEquals(1.0, Expression.eval("log(${E})").asNumber(), 1e-10)
        assertEquals(1.0, Expression.eval("log10(10)").asNumber(), 1e-10)
        
        // Test error cases
        assertThrows<ExpressionError> {
            Expression.eval("log(-1)")
        }
        assertThrows<ExpressionError> {
            Expression.eval("log10(0)")
        }
    }
    
    @Test
    fun testCustomFunctions() {
        env.setFunction("double") { args ->
            if (args.size != 1) throw ExpressionError("double() requires exactly 1 argument")
            Value.number(args[0].asNumber() * 2)
        }
        
        env.setFunction("concat") { args ->
            if (args.size != 2) throw ExpressionError("concat() requires exactly 2 arguments")
            Value.string(args[0].asString() + args[1].asString())
        }
        
        assertEquals(Value.number(10.0), Expression.eval("double(5)", env))
        assertEquals(Value.string("helloworld"), Expression.eval("concat(\"hello\", \"world\")", env))
    }
    
    @Test
    fun testFunctionNotFound() {
        val exception = assertThrows<ExpressionError> {
            Expression.eval("unknownFunc(1, 2)")
        }
        assertTrue(exception.message!!.contains("Function 'unknownFunc' not found"))
    }
    
    // MARK: - String Literal Tests
    
    @Test
    fun testStringLiterals() {
        assertEquals(Value.string("hello"), Expression.eval("\"hello\""))
        assertEquals(Value.string(""), Expression.eval("\"\""))
        assertEquals(Value.string("hello world"), Expression.eval("\"hello world\""))
    }
    
    @Test
    fun testStringConcatenation() {
        assertEquals(Value.string("helloworld"), Expression.eval("\"hello\" + \"world\""))
        assertEquals(Value.string("hello5"), Expression.eval("\"hello\" + 5"))
        assertEquals(Value.string("5hello"), Expression.eval("5 + \"hello\""))
        assertEquals(Value.string("hellotrue"), Expression.eval("\"hello\" + true"))
    }
    
    @Test
    fun testStringEscapes() {
        assertEquals(Value.string("hello\nworld"), Expression.eval("\"hello\\nworld\""))
        assertEquals(Value.string("hello\tworld"), Expression.eval("\"hello\\tworld\""))
        assertEquals(Value.string("hello\"world"), Expression.eval("\"hello\\\"world\""))
        assertEquals(Value.string("hello\\world"), Expression.eval("\"hello\\\\world\""))
    }
    
    // MARK: - Type Conversion Tests
    
    @Test
    fun testNumberToBoolean() {
        assertTrue(Value.number(1.0).asBoolean())
        assertTrue(Value.number(-1.0).asBoolean())
        assertTrue(Value.number(0.1).asBoolean())
        assertFalse(Value.number(0.0).asBoolean())
    }
    
    @Test
    fun testBooleanToNumber() {
        assertEquals(1.0, Value.boolean(true).asNumber())
        assertEquals(0.0, Value.boolean(false).asNumber())
    }
    
    @Test
    fun testStringToNumber() {
        assertEquals(42.0, Value.string("42").asNumber())
        assertEquals(3.14, Value.string("3.14").asNumber())
        assertEquals(-5.0, Value.string("-5").asNumber())
        
        assertThrows<ExpressionError> {
            Value.string("not a number").asNumber()
        }
    }
    
    @Test
    fun testStringToBoolean() {
        // Empty string is false
        assertFalse(Value.string("").asBoolean())
        
        // Explicit false values
        assertFalse(Value.string("false").asBoolean())
        assertFalse(Value.string("no").asBoolean())
        assertFalse(Value.string("0").asBoolean())
        
        // All other non-empty strings are true
        assertTrue(Value.string("true").asBoolean())
        assertTrue(Value.string("yes").asBoolean())
        assertTrue(Value.string("hello").asBoolean())
        assertTrue(Value.string("1").asBoolean())
    }
    
    @Test
    fun testValueToString() {
        assertEquals("42", Value.number(42.0).asString())
        assertEquals("true", Value.boolean(true).asString())
        assertEquals("false", Value.boolean(false).asString())
        assertEquals("hello", Value.string("hello").asString())
    }
    
    // MARK: - Error Handling Tests
    
    @Test
    fun testDivisionByZero() {
        assertThrows<ExpressionError> {
            Expression.eval("5 / 0")
        }
    }
    
    @Test
    fun testModuloByZero() {
        assertThrows<ExpressionError> {
            Expression.eval("5 % 0")
        }
    }
    
    @Test
    fun testSyntaxErrors() {
        // Missing closing parenthesis
        assertThrows<ExpressionError> {
            Expression.eval("(2 + 3")
        }
        
        // Missing opening parenthesis
        assertThrows<ExpressionError> {
            Expression.eval("2 + 3)")
        }
        
        // Invalid function call syntax
        assertThrows<ExpressionError> {
            Expression.eval("sqrt(")
        }
        
        // Unexpected character
        assertThrows<ExpressionError> {
            Expression.eval("2 + @ + 3")
        }
    }
    
    @Test
    fun testInvalidFunctionArguments() {
        // sqrt with wrong number of arguments
        assertThrows<ExpressionError> {
            Expression.eval("sqrt(1, 2)")
        }
        
        // sqrt of negative number
        assertThrows<ExpressionError> {
            Expression.eval("sqrt(-1)")
        }
    }
    
    // MARK: - Token Collection Tests
    
    @Test
    fun testTokenCollection() {
        env.setVariable("x", 5.0)
        val (result, tokens) = Expression.eval("2 + 3 * x", env, true)
        
        assertTrue(tokens.isNotEmpty())
        
        // Should contain number, operator, number, operator, identifier tokens
        val tokenTypes = tokens.map { it.type }
        assertTrue(tokenTypes.contains(TokenType.NUMBER), "Should contain NUMBER token")
        assertTrue(tokenTypes.contains(TokenType.OPERATOR), "Should contain OPERATOR token") 
        assertTrue(tokenTypes.contains(TokenType.IDENTIFIER), "Should contain IDENTIFIER token")
    }
    
    @Test
    fun testTokenPositions() {
        env.setVariable("x", 5.0)
        val (_, tokens) = Expression.eval("42 + sqrt(x)", env, true)
        
        // Verify token positions are sequential and non-overlapping
        var lastEnd = 0
        for (token in tokens.filter { it.type != TokenType.WHITESPACE }) {
            assertTrue(token.start >= lastEnd, "Token positions should be sequential")
            lastEnd = token.start + token.length
        }
    }
    
    // MARK: - Complex Expression Tests
    
    @Test
    fun testComplexExpressions() {
        env.setVariable("a", 10.0)
        env.setVariable("b", 5.0)
        env.setVariable("c", 2.0)
        
        // Complex arithmetic with variables
        assertEquals(Value.number(30.0), Expression.eval("a + b * c + a", env))
        
        // Mixed boolean and arithmetic
        assertEquals(Value.boolean(true), Expression.eval("a > b && b > c", env))
        
        // Function calls in complex expressions
        assertEquals(Value.number(sqrt(15.0) * sqrt(4.0)), Expression.eval("sqrt(a + b) * sqrt(b - 1)", env))
    }
    
    @Test
    fun testNestedFunctionCalls() {
        assertEquals(Value.number(3.0), Expression.eval("sqrt(sqrt(81))"))
        assertEquals(Value.number(16.0), Expression.eval("pow(sqrt(16), 2)"))
        assertEquals(Value.number(8.0), Expression.eval("max(min(10, 8), 6)"))
    }
    
    // MARK: - Parse and Evaluate Separately Tests
    
    @Test
    fun testParseAndEvaluate() {
        env.setVariable("x", 5.0)
        env.setVariable("y", 3.0)
        
        // Parse expression once
        val ast = Expression.parse("x * 2 + y")
        
        // Evaluate multiple times with different variable values
        assertEquals(Value.number(13.0), ast.evaluate(env))
        
        env.setVariable("x", 10.0)
        assertEquals(Value.number(23.0), ast.evaluate(env))
        
        env.setVariable("y", 7.0)
        assertEquals(Value.number(27.0), ast.evaluate(env))
    }
    
    // MARK: - Edge Cases
    
    @Test
    fun testEmptyExpression() {
        assertThrows<ExpressionError> {
            Expression.eval("")
        }
    }
    
    @Test
    fun testWhitespaceOnly() {
        assertThrows<ExpressionError> {
            Expression.eval("   \t\n  ")
        }
    }
    
    @Test
    fun testUnaryOperators() {
        assertEquals(Value.number(5.0), Expression.eval("+5"))
        assertEquals(Value.number(-5.0), Expression.eval("-5"))
        assertEquals(Value.number(5.0), Expression.eval("-(-5)"))
        assertEquals(Value.boolean(false), Expression.eval("!true"))
        assertEquals(Value.boolean(true), Expression.eval("!!true"))
    }
}